app/anti_virus_check.py

```diff
@@ -0,0 +1,136 @@
+import collections
+import time
+
+import requests
+from sdc.rabbit.exceptions import QuarantinableError, RetryableError, BadMessageError
+
+from app import create_and_wrap_logger
+from app import settings
+
+logger = create_and_wrap_logger(__name__)
+
+AVResult = collections.namedtuple('AVResult', 'safe ready scan_results')
+
+
+class AntiVirusCheck:
+    def __init__(self, tx_id):
+        self.bound_logger = logger.bind(tx_id=tx_id)
+
+    def send_for_av_scan(self, payload):
+        self.bound_logger.info("Sending for AV check", filename=payload.file_name)
+        data_id = self._send_for_anti_virus_check(payload.file_name, payload.decoded_contents)
+        self.bound_logger.info("Sent for A/V check", data_id=data_id)
+
+        attempts = 0
+        while attempts <= settings.ANTI_VIRUS_MAX_ATTEMPTS:
+            attempts += 1
+            results = self._get_anti_virus_result(data_id)
+            if not results.ready:
+                time.sleep(settings.ANTI_VIRUS_WAIT_TIME)
+            elif not results.safe:
+                error = "Unsafe file detected for case id {} with filename {}".format(payload.case_id,
+                                                                                      payload.file_name)
+                self._write_scan_report(results, payload.file_name)
+                self.bound_logger.error(error)
+                raise QuarantinableError()
+            else:
+                self.bound_logger.debug(
+                    "File {} for case id {} has been virus checked and confirmed safe".format(payload.case_id,
+                                                                                              payload.file_name))
+                return True
+
+        # out of attempts raise retryable error to force the response back to the queue.
+        self.bound_logger.error(
+            "Unable to get results of Anti-virus scan for case id {} and file {} ".format(payload.case_id,
+                                                                                          payload.file_name))
+        raise RetryableError()
+
+    def _add_api_key(self, headers):
+        # the API key is only needed for the online cloud service
+        if settings.ANTI_VIRUS_API_KEY:
+            self.bound_logger.debug("Setting A/V API key")
+            headers['apikey'] = settings.ANTI_VIRUS_API_KEY
+
+    def _check_av_response(self, response):
+        if response.status_code != 200:
```

Comment:
This would be far more readable using [`response.raise_for_status()`](http://docs.python-requests.org/en/master/api/#requests.Response.raise_for_status)

Reply:
agreed